File: lrpc-csharp/src/lrpc/val/Store.cs
using System.Text;

namespace lrpc.val
{
    /// <summary>
    /// rules for storing and restoring data
    /// </summary>
    public interface Store
    {
        /// <summary>
        /// storing data
        /// </summary>
        /// <param name="que">queue for storing data</param>
        /// <param name="val">the data to be stored is not a class that implements this interface</param>
        void Store(ByteQue que, object val);
        /// <summary>
        /// restoring data
        /// </summary>
        /// <param name="que">queue for restoring data</param>
        /// <returns>previously stored data</returns>
        object Restore(ByteQue que);
    }

    class ByteStore : Store
    {
        public void Store(ByteQue que, object val)
        {
            if (val == null)
            {
                que.Push(false);
            }
            else
            {
                que.Push(true);
                que.Push((byte)val);
            }
        }

        public object Restore(ByteQue que)
        {
            if (que.Pop<bool>())
            {
                return que.Pop<byte>();
            }
            return null;
        }
    }

    class SByteStore : Store
    {
        public void Store(ByteQue que, object val)
        {
            if (val == null)
            {
                que.Push(false);
            }
            else
            {
                que.Push(true);
                que.Push((sbyte)val);
            }
        }

        public object Restore(ByteQue que)
        {
            if (que.Pop<bool>())
            {
                return que.Pop<sbyte>();
            }
            return null;
        }
    }

    class ShortStore : Store
    {
        public void Store(ByteQue que, object val)
        {
            if (val == null)
            {
                que.Push(false);
            }
            else
            {
                que.Push(true);
                que.Push((short)val);
            }
        }

        public object Restore(ByteQue que)
        {
            if (que.Pop<bool>())
            {
                return que.Pop<short>();
            }
            return null;
        }
    }

    class UShortStore : Store
    {
        public void Store(ByteQue que, object val)
        {
            if (val == null)
            {
                que.Push(false);
            }
            else
            {
                que.Push(true);
                que.Push((ushort)val);
            }
        }

        public object Restore(ByteQue que)
        {
            if (que.Pop<bool>())
            {
                return que.Pop<ushort>();
            }
            return null;
        }
    }

    class IntStore : Store
    {
        public void Store(ByteQue que, object val)
        {
            if (val == null)
            {
                que.Push(false);
            }
            else
            {
                que.Push(true);
                que.Push((int)val);
            }
        }

        public object Restore(ByteQue que)
        {
            if (que.Pop<bool>())
            {
                return que.Pop<int>();
            }
            return null;
        }
    }

    class UIntStore : Store
    {
        public void Store(ByteQue que, object val)
        {
            if (val == null)
            {
                que.Push(false);
            }
            else
            {
                que.Push(true);
                que.Push((uint)val);
            }
        }

        public object Restore(ByteQue que)
        {
            if (que.Pop<bool>())
            {
                return que.Pop<uint>();
            }
            return null;
        }
    }

    class LongStore : Store
    {
        public void Store(ByteQue que, object val)
        {
            if (val == null)
            {
                que.Push(false);
            }
            else
            {
                que.Push(true);
                que.Push((long)val);
            }
        }

        public object Restore(ByteQue que)
        {
            if (que.Pop<bool>())
            {
                return que.Pop<long>();
            }
            return null;
        }
    }

    class ULongStore : Store
    {
        public void Store(ByteQue que, object val)
        {
            if (val == null)
            {
                que.Push(false);
            }
            else
            {
                que.Push(true);
                que.Push((ulong)val);
            }
        }

        public object Restore(ByteQue que)
        {
            if (que.Pop<bool>())
            {
                return que.Pop<ulong>();
            }
            return null;
        }
    }

    class FloatStore : Store
    {
        public void Store(ByteQue que, object val)
        {
            if (val == null)
            {
                que.Push(false);
            }
            else
            {
                que.Push(true);
                que.Push((float)val);
            }
        }

        public object Restore(ByteQue que)
        {
            if (que.Pop<bool>())
            {
                return que.Pop<float>();
            }
            return null;
        }
    }

    class DoubleStore : Store
    {
        public void Store(ByteQue que, object val)
        {
            if (val == null)
            {
                que.Push(false);
            }
            else
            {
                que.Push(true);
                que.Push((double)val);
            }
        }

        public object Restore(ByteQue que)
        {
            if (que.Pop<bool>())
            {
                return que.Pop<double>();
            }
            return null;
        }
    }

    class BoolStore : Store
    {
        public void Store(ByteQue que, object val)
        {
            if (val == null)
            {
                que.Push(false);
            }
            else
            {
                que.Push(true);
                que.Push((bool)val);
            }
        }

        public object Restore(ByteQue que)
        {
            if (que.Pop<bool>())
            {
                return que.Pop<bool>();
            }
            return null;
        }
    }

    class CharStore : Store
    {
        public void Store(ByteQue que, object val)
        {
            if (val == null)
            {
                que.Push(false);
            }
            else
            {
                que.Push(true);
                que.Push((char)val);
            }
        }

        public object Restore(ByteQue que)
        {
            if (que.Pop<bool>())
            {
                return que.Pop<char>();
            }
            return null;
        }
    }

    class StringStore : Store
    {
        public void Store(ByteQue que, object val)
        {
            if (val == null)
            {
                que.Push(false);
            }
            else
            {
                que.Push(true);
                byte[] arr = Encoding.UTF8.GetBytes((string)val);
                que.PushSize(arr.Length);
                foreach (byte ch in arr)
                {
                    que.Push(ch);
                }
            }
        }

        public object Restore(ByteQue que)
        {
            if (que.Pop<bool>())
            {
                byte[] arr = new byte[que.PopSize()];
                for (int i = 0; i < arr.Length; ++i)
                {
                    arr[i] = que.Pop<byte>();
                }
                return Encoding.UTF8.GetString(arr);
            }
            return null;
        }
    }
}

File: lrpc-java/src/com/lrpc/val/Store.java
package com.lrpc.val;

import java.nio.charset.Charset;

/**
 * rules for storing and restoring data
 */
public interface Store {

    /**
     * storing data
     * 
     * @param que queue for storing data
     * @param val the data to be stored is not a class that implements this
     *            interface
     */
    void store(ByteQue que, Object val);

    /**
     * restoring data
     * 
     * @param que queue for restoring data
     * @return previously stored data
     */
    Object restore(ByteQue que);
}

class ByteStore implements Store {

    @Override
    public void store(ByteQue que, Object val) {
        if (val == null) {
            que.push(boolean.class, false);
        } else {
            que.push(boolean.class, true);
            que.push(byte.class, val);
        }
    }

    @Override
    public Object restore(ByteQue que) {
        if ((boolean) que.pop(boolean.class)) {
            return que.pop(byte.class);
        }
        return null;
    }
}

class ShortStore implements Store {

    @Override
    public void store(ByteQue que, Object val) {
        if (val == null) {
            que.push(boolean.class, false);
        } else {
            que.push(boolean.class, true);
            que.push(short.class, val);
        }
    }

    @Override
    public Object restore(ByteQue que) {
        if ((boolean) que.pop(boolean.class)) {
            return que.pop(short.class);
        }
        return null;
    }
}

class IntegerStore implements Store {

    @Override
    public void store(ByteQue que, Object val) {
        if (val == null) {
            que.push(boolean.class, false);
        } else {
            que.push(boolean.class, true);
            que.push(int.class, val);
        }
    }

    @Override
    public Object restore(ByteQue que) {
        if ((boolean) que.pop(boolean.class)) {
            return que.pop(int.class);
        }
        return null;
    }
}

class LongStore implements Store {

    @Override
    public void store(ByteQue que, Object val) {
        if (val == null) {
            que.push(boolean.class, false);
        } else {
            que.push(boolean.class, true);
            que.push(long.class, val);
        }
    }

    @Override
    public Object restore(ByteQue que) {
        if ((boolean) que.pop(boolean.class)) {
            return que.pop(long.class);
        }
        return null;
    }
}

class FloatStore implements Store {

    @Override
    public void store(ByteQue que, Object val) {
        if (val == null) {
            que.push(boolean.class, false);
        } else {
            que.push(boolean.class, true);
            que.push(float.class, val);
        }
    }

    @Override
    public Object restore(ByteQue que) {
        if ((boolean) que.pop(boolean.class)) {
            return que.pop(float.class);
        }
        return null;
    }
}

class DoubleStore implements Store {

    @Override
    public void store(ByteQue que, Object val) {
        if (val == null) {
            que.push(boolean.class, false);
        } else {
            que.push(boolean.class, true);
            que.push(double.class, val);
        }
    }

    @Override
    public Object restore(ByteQue que) {
        if ((boolean) que.pop(boolean.class)) {
            return que.pop(double.class);
        }
        return null;
    }
}

class BooleanStore implements Store {

    @Override
    public void store(ByteQue que, Object val) {
        if (val == null) {
            que.push(boolean.class, false);
        } else {
            que.push(boolean.class, true);
            que.push(boolean.class, val);
        }
    }

    @Override
    public Object restore(ByteQue que) {
        if ((boolean) que.pop(boolean.class)) {
            return que.pop(boolean.class);
        }
        return null;
    }
}

class CharacterStore implements Store {

    @Override
    public void store(ByteQue que, Object val) {
        if (val == null) {
            que.push(boolean.class, false);
        } else {
            que.push(boolean.class, true);
            que.push(char.class, val);
        }
    }

    @Override
    public Object restore(ByteQue que) {
        if ((boolean) que.pop(boolean.class)) {
            return que.pop(char.class);
        }
        return null;
    }
}

class StringStore implements Store {

    @Override
    public void store(ByteQue que, Object val) {
        if (val == null) {
            que.push(boolean.class, false);
        } else {
            que.push(boolean.class, true);
            byte[] arr = ((String) val).getBytes(Charset.forName("UTF-8"));
            que.pushSize(arr.length);
            for (byte ch : arr) {
                que.push(byte.class, ch);
            }
        }
    }

    @Override
    public Object restore(ByteQue que) {
        if ((boolean) que.pop(boolean.class)) {
            byte[] arr = new byte[que.popSize()];
            for (int i = 0; i < arr.length; ++i) {
                arr[i] = (byte) que.pop(byte.class);
            }
            return new String(arr, Charset.forName("UTF-8"));
        }
        return null;
    }
}